packages/zoe/src/zoeService/installationStorage.js

```diff
@@ -80,54 +72,79 @@ export const makeInstallationStorage = (
     return installation;
   };
 
-  /** @type {InstallBundle} */
-  const installBundle = async allegedBundle => {
-    // Bundle is a very open-ended type and we must decide here
-    // whether to treat it as either a HashBundle or SourceBundle.
-    // So, we cast it down and inspect it.
-    const bundle = /** @type {unknown} */ (harden(allegedBundle));
-    assert.typeof(bundle, 'object', 'a bundle must be provided');
-    assert(bundle !== null, 'a bundle must be provided');
-    const { moduleFormat } = bundle;
-    if (moduleFormat === 'endoZipBase64Sha512') {
-      const { endoZipBase64Sha512 } = bundle;
-      assert.typeof(
-        endoZipBase64Sha512,
-        'string',
-        `bundle endoZipBase64Sha512 must be a string, got ${endoZipBase64Sha512}`,
-      );
-      return installBundleID(`b1-${endoZipBase64Sha512}`);
-    }
-    return installSourceBundle(bundle);
-  };
+  const InstallationStorage = M.interface('InstallationStorage', {
+    installBundle: M.call(BundleShape).returns(M.promise()),
+    installBundleID: M.call(M.string()).returns(M.promise()),
+    unwrapInstallation: M.callWhen(M.eref(InstallationShape)).returns(M.any()),
```

Comment:
A wonderful opportunity to show off the power of `callWhen`. By changing this to

```suggestion
    unwrapInstallation: M.callWhen(M.await(InstallationShape)).returns(M.any()),
```

the raw method below will not be called unless and until
   * the incoming argument, if it is a promise, fulfills to something that passes `InstallationShape`
   * if the incoming argument is not a promise, it is also validated against `InstallationShape`

Only after we have an argument that is validated against `InstallationShape` is the raw method below called with that argument, so it doesn't need to do that job itself.

In the promise case, this is better than the current PR. In the current PR, the promise will pass through the parameter pattern no matter what it resolves to, and is then passed to the raw method below, which does the `E.when` dance in case it is a promise. But it never checks the fulfillment against `InstallationShape` either. So await does the input validation that is otherwise easily missed.

Reply:
done